### PR TITLE
[QSTR-116] Adding logic to execute any query that is sent to QStore.

### PR DESCRIPTION
**Pull request by suprajkolluri**
Adding logic to execute any query that is sent to QStore and return the result

### CODE REVIEW

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 1]
```diff
@@ -43,67 +45,74 @@
     @Autowired
     IRepositoryManager repositorymanager;
 
+    @Autowired
+    @Qualifier("neo4jOperations")
+    private Neo4jOperations template;
+
+    @Autowired
+    Converter converter;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: where is the access modifier?

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 2]
```diff
@@ -194,6 +196,38 @@ public String getData(HttpServletRequest request,
 
     }
 
+    @ResponseBody
+    @RequestMapping(value = "/getResultForQuery", method = RequestMethod.POST)
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: bad api method name. why not just ```query```?

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 3]
```diff
@@ -194,6 +196,38 @@ public String getData(HttpServletRequest request,
...
+    public String getAllNetworks(HttpServletRequest request, HttpServletResponse response,
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: bad method name

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 4]
```diff
@@ -194,6 +196,38 @@ public String getData(HttpServletRequest request,
...
+            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas)
+                    throws JSONException {
+
+        if (query.isEmpty()) {
+            response.setStatus(400);
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: you should use the ```httpstatus``` enum here

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 5]
```diff
@@ -194,6 +196,38 @@ public String getData(HttpServletRequest request,
...
+            return "Please provide the query to execute the request";
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: method should return valid xml/json

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 6]
```diff
@@ -194,6 +196,38 @@ public String getData(HttpServletRequest request,
...
+        }
+
+        if (clas.isEmpty()) {
+            response.setStatus(400);
+            return "Please provide the Class to execute the request";
+        }
+
+        logger.info("Requested format: " + accept);
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: why does this need to be logged?

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 7]
```diff
@@ -43,67 +45,74 @@
     @Autowired
     IRepositoryManager repositorymanager;
 
+    @Autowired
+    @Qualifier("neo4jOperations")
+    private Neo4jOperations template;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: do you actually use that object?

File: QStore4S/src/main/java/edu/asu/qstore4s/db/neo4j/impl/DbConnector.java [hunk 8]
```diff
@@ -850,8 +670,30 @@ private String buildWhereClause(String name, String property, String value,
...
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CreationEvent> executeQuery(String query, String clas) {
+        Class<?> classs = null;
+        if ("relationevent".equals(clas.toLowerCase())) {
+            classs = RelationEvent.class;
+        } else {
+            classs = AppellationEvent.class;
+        }
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/db/neo4j/impl/DbConnector.java]: this is a bad approach. What if we want to return other objects in the future? This method should take a class object, and then you should validate the input string much earlier and turn it into a class object.

File: QStore4S/src/main/java/edu/asu/qstore4s/db/neo4j/impl/DbConnector.java [hunk 9]
```diff
@@ -850,8 +670,30 @@ private String buildWhereClause(String name, String property, String value,
...
+        return "( HAS (" + name + "." + property + ") and " + name + "." + property + "='" + value + "') ";
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CreationEvent> executeQuery(String query, String clas) {
+        Class<?> classs = null;
+        if ("relationevent".equals(clas.toLowerCase())) {
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/db/neo4j/impl/DbConnector.java]: (almost) never hard-code strings...!

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 10]
```diff
@@ -36,21 +35,20 @@
...
-    @Autowired
-    IRepositoryManager repositorymanager;
-
-    @Autowired
-    @Qualifier("neo4jOperations")
-    private Neo4jOperations template;
+    private static final HashMap<String, Class<?>> classMap = new HashMap<String, Class<?>>() {
+        {
+            put("relationevent", RelationEvent.class);
+            put("appellationevent", AppellationEvent.class);
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: these strings should be constants.

Comment (jdamerow) [reply QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: should also not be static. There is simply no reason here why that would have to be static, and if anything only makes it more difficult to maintain.

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 11]
```diff
@@ -36,21 +35,20 @@
...
+        }
+    };
 
     @Autowired
-    Converter converter;
+    IRepositoryManager repositorymanager;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: access modifier missing

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 12]
```diff
@@ -189,6 +190,41 @@ public String getData(HttpServletRequest request,
...
+    }
+
+    @ResponseBody
+    @RequestMapping(value = "/query", method = RequestMethod.POST)
+    public String query(HttpServletRequest request, HttpServletResponse response,
+            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas)
+                    throws JSONException {
+
+        edu.asu.qstore4s.error.Error error = null;
+        if (query.isEmpty()) {
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java:206]: should probably be trimmed

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 13]
```diff
@@ -189,6 +190,41 @@ public String getData(HttpServletRequest request,
...
+
+    }
+
+    @ResponseBody
+    @RequestMapping(value = "/query", method = RequestMethod.POST)
+    public String query(HttpServletRequest request, HttpServletResponse response,
+            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas)
+                    throws JSONException {
+
+        edu.asu.qstore4s.error.Error error = null;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: initialize where needed

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 14]
```diff
@@ -189,6 +190,41 @@ public String getData(HttpServletRequest request,
...
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            error = new Error("Please provide the query to execute the request");
+            return error.toString(accept);
+        }
+
+        Class<?> classs = classMap.get(clas);
+        if (classs == null) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            error = new Error("Please provide a valid Class to execute the request");
+            return error.toString(accept);
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: you can probably also merge the two lines into one

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 15]
```diff
@@ -251,31 +283,28 @@ public String getData(HttpServletRequest request,
...
-            IOException, URISyntaxException, ParseException, JSONException,
-            InvalidDataException {
+    public String searchByAppellationId(HttpServletRequest request, HttpServletResponse response,
+            @RequestBody String xml, @RequestHeader("Accept") String accept) throws ParserException, IOException,
+                    URISyntaxException, ParseException, JSONException, InvalidDataException {
 
-        logger.debug("Requested format: " +  accept);
         String trimmedXML = xml.trim();
+
+        edu.asu.qstore4s.error.Error error = null;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: initialize where needed

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 16]
```diff
@@ -300,18 +329,17 @@ public String searchByAppellationId(HttpServletRequest request,
...
+            @RequestHeader("Accept") String accept) throws ParserException, IOException, URISyntaxException,
+                    ParseException, JSONException, InvalidDataException {
 
-        logger.debug("Requested format: " +  accept);
         String trimmedXML = xml.trim();
-        if (trimmedXML.isEmpty()) {
-            response.setStatus(500);
-            return "Please provide XML in body of the post request.";
 
+        edu.asu.qstore4s.error.Error error = null;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: same here

File: QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/AppellationEvent.java [hunk 17]
```diff
@@ -12,17 +13,28 @@
...
-		return term;
-	}
-
-	public void setTerm(Term term) {
-		this.term = term;
-	}
+public class AppellationEvent extends CreationEvent {
+
+    @GraphId
+    Long graphId;
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/AppellationEvent.java]: why don't you use the creationEvent graph id?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java [hunk 18]
```diff
@@ -189,6 +190,39 @@ public String getData(HttpServletRequest request,
...
+            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas)
+                    throws JSONException {
+
+        query = query.trim();
+        if (query.isEmpty()) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            return new Error("Please provide the query to execute the request").toString(accept);
+        }
+
+        Class<?> classs = classMap.get(clas);
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java]: typically class objects are called ```clazz```. not a big deal, but ```classs``` looks odd.

File: QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/CreationEvent.java [hunk 19]
```diff
@@ -29,8 +29,7 @@
 	@GraphId
 	Long graphId;
 	
-
-	@Property(name = "source_reference")
+    @Property(name = "source_reference")
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/CreationEvent.java:32]: formatting is off

File: QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/CreationEvent.java [hunk 20]
```diff
@@ -69,5 +68,13 @@ public Actor getInterpretationCreator() {
...
 		this.interpretation_creator = interpretationCreator;
 	}
+	
+	public Long getGraphId() {
+        return graphId;
+    }
+
+    public void setGraphId(Long graphId) {
+        this.graphId = graphId;
+    }
```
Comment (jdamerow) [inline QStore4S/src/main/java/edu/asu/qstore4s/domain/events/impl/CreationEvent.java:78]: here as well

**Comment by jdamerow** [comment]
once formatting has been fixed should be ready for merging

**Review by jdamerow — APPROVED** [review]

**Comment by suprajkolluri** [comment]
@jdamerow Fixed the variable name and formatting

**Pull request merged**